src/pepr/operator/controllers/keycloak/authservice/authservice.ts

```diff
@@ -0,0 +1,127 @@
+import { createHash } from "crypto";
+import { K8s, Log, kind } from "pepr";
+import { UDSConfig } from "../../../../config";
+import { UDSPackage } from "../../../crd";
+import { updatePolicy } from "./authorization-policy";
+import { Action, AuthServiceEvent, AuthserviceConfig, Chain } from "./types";
+
+const namespace = "authservice";
+const secretName = "authservice";
+const baseDomain = `https://sso.${UDSConfig.domain}`;
+const realm = "uds";
+
+export async function reconcileAuthservice(
+  event: AuthServiceEvent,
+  labelSelector: { [key: string]: string },
+  pkg: UDSPackage,
+) {
+  await updateConfig(event);
+  await updatePolicy(event, labelSelector, pkg);
+}
+
+// write authservice config to secret
+export async function updateConfig(event: AuthServiceEvent) {
+  const authSvcSecret = await K8s(kind.Secret).InNamespace(namespace).Get(secretName);
+
+  // parse existing authservice config
+  let config = JSON.parse(atob(authSvcSecret!.data!["config.json"])) as AuthserviceConfig;
+
+  // update config based on event
+  config = buildConfig(config, event);
+
+  // update the authservice secret
+  await updateAuthServiceSecret(config);
+}
+
+// might be neccesary to rebuild the entire config
+export function buildConfig(config: AuthserviceConfig, event: AuthServiceEvent) {
+  let chains: Chain[];
+
+  if (event.action == Action.Add) {
+    // add the new chain to the existing authservice config
+    chains = config.chains.concat(buildChain(event));
+  } else {
```

Comment:
A side effect of this logic and the way we `Reconcile` `Package`s on pepr restart - everytime I cycle the watcher pod I end up duplicating the chain in the authservice secret. We likely need to handle some logic around modify/de-dupe for the chains here.